### ISSUE TITLE
Compatibility with React Native New Architecture

### ISSUE THREAD

**Issue by FlavioScimeca** [transistorsoft/react-native-background-geolocation#2210]
- Plugin version: 
- Platform: Both
- OS version: 
- Device manufacturer / model:
- React Native version: 0.76.0

I currently have a production-ready app using react-native-background-geolocation. I am considering upgrading the project to use React Native’s new architecture, but the documentation does not explicitly confirm whether the library is fully compatible. Before proceeding, I need to ensure there are no compatibility issues or significant steps required for integration.

**Comment by christocracy** [comment]
I’m not aware of any way that could affect the plug-in.  The plug-in operates largely unaffected by React Native.  Why don’t you try it and see.

**Comment by FlavioScimeca** [comment]
I asked because I should update the other dependencies as well , anyway thank you

**Comment by christocracy** [comment]
The plug-in is definitely not affect by anything related to UI or JavaScript.

this plug-in exists mostly within pure native libraries that don’t even know that RN exists.  These libraries are common to all versions of the plug-in (Cordova, Capacitor, Flutter, React Native).  These native libs can even work in pure native apps.

there’s not much that RN can do to affect this sdk.

**Comment by christocracy** [comment]
React Native seems to have an issue with running "Headless" code with *new archictecture*.  Both this plugin and react-native-background-fetch make use of react-native's "HeadlessJS" mechanism.  A reproduction example has been provided to the React Native team two days ago.

See https://github.com/facebook/react-native/issues/47592

**Comment by mikehardy** [comment]
@christocracy react-native can quietly alter the android modular teardown callback from `onCatalystInstanceDestroy` to `invalidate` and then leave `onCatalystInstanceDestroy` in there so that it doesn't cause a build failure but nothing ever calls it - that one bit me in react-native-firebase, fixed it just a couple days back 😆 

For react-native 0.76.2 (current as of this typing) I do not see any problems with headlessJS anymore on new architecture (which is now the default in 0.76.0+) - contrary to your statement above. 0.76.2+new-arch should work.

I *do* see problems now with *old architecture* on react-native >= 0.76.0 through at least 0.76.2, and I logged an upstream reproducer for that in the issue you linked.

**Comment by christocracy** [comment]
`react-native-background-fetch` has been released with support for *new architecture (bridgeless)*.  Next up will be `background-geolocation` (working but not yet released).

- It seems to work fine on `react-native v0.76.3` with both `newArchEnabled=true` and `newArchEnabled=false`.
- I tested it on an older version of `react-native v0.70.6`.

@mikehardy I copied your idea of using reflection.  See [`HeadlessTask.java`](https://github.com/transistorsoft/react-native-background-fetch/blob/master/android/src/main/java/com/transistorsoft/rnbackgroundfetch/HeadlessTask.java)

Also, I just discovered [`consumerProguardFiles`](https://github.com/transistorsoft/react-native-background-fetch/blob/a40287b2753fa99640067ebfe5d93501397b7b17/android/build.gradle#L23).  I wasn't aware that an Android library could provide its own `proguard-rules` to the host application.  That allows me to remove a setup step!

Here's my [`proguard-rules`](https://github.com/transistorsoft/react-native-background-fetch/blob/master/android/proguard-rules.pro).  I added `-keep class com.facebook.react.fabric.** { *; }` but I think that's to fix an issue I had with my demo app and `react-navigation`.

**Comment by mikehardy** [comment]
nice one - read through that file you've got a lot more going on with ReactContext start listeners, UI thread assertions, and posting the task and stuff 🤔  is that because you aren't extending the upstream base headlessJS - neither notifee nor react-native-firebase are being as careful as your code is ?

**Comment by christocracy** [comment]
> is that because you aren't extending the upstream base headlessJS

I don’t know.  I’ve always based my `HeadlessTask` implementation upon what I find in RN’s [`HeadlessJsTaskService`](https://github.com/facebook/react-native/blob/main/packages/react-native/ReactAndroid/src/main/java/com/facebook/react/HeadlessJsTaskService.java), without the `WakeLock` stuff.  

I’m unfamiliar with this “*upstream base headlessJS*”.

**Comment by mikehardy** [comment]
> I’m unfamiliar with this “upstream base headlessJS”.

https://github.com/invertase/react-native-firebase/blob/main/packages/messaging/android/src/main/java/io/invertase/firebase/messaging/ReactNativeFirebaseMessagingHeadlessService.java#L11

If you extend it, you just need to implement getTaskConfig, I think? But your implementation makes me want to look more deeply if I ever have high priority stuff cleared away...

**Comment by christocracy** [comment]
Right, so you're *extending* the `HeadlessJsTaskService`, where I just plucked the code I wanted from it.

I don't want my `HeadlessTask` to launch a service.

- For `background-fetch`, the event is fired from a `JobScheduler` event, which is already providing background-processing time-- there's no need to launch a `Service` here.
- For `background-geolocation`, the headless event is most often invoked from one of its already-running foreground-services.  If the developer is going to run a long-running task, they can run it within `BackgroundGeolocation.startBackgroundTask'

**Comment by mikehardy** [comment]
that makes sense, and also explains why Notifee has a more similar style to the one here but I couldn't quite square it with the minimal implementation in react-native-firebase - Notifee may be waking up from WorkManager with time-based scheduled notifications. This area is a bit new to me so the subtleties are all novel, thanks for the extra detail

**Comment by christocracy** [comment]
You're aware of the complexity of launching a `Service` in the background, of course:  the requirement of the foreground-service `Notification`, calling `Context.startForeground` within 10 seconds, and all that jazz.  RN's `HeadlessJsTaskService` doesn't implement any of that.

**Comment by vonkanehoffen** [comment]
So... was this resolved then @christocracy ? Are headless tasks and all that good stuff that `react-native-background-geolocation` does good to go for New Architecture?

**Comment by christocracy** [comment]
When in doubt, see the [CHANGELOG](https://github.com/transistorsoft/react-native-background-geolocation/blob/master/CHANGELOG.md)

![Image](https://github.com/user-attachments/assets/7f5636ba-0f2a-4674-9142-f57a77ec10f3)
